// This file describes common exchange format for tarc document data.
// Draft. Some data from text archive may be missing.

package NTextArc;

option optimize_for = SPEED;

// Describes document sentence in terms of offsets to main doc body
message TDocSentence {
    // Offset in TDocArcData::Text to first byte of the sentence
    optional uint32 Offset = 1;

    // Size of sentence in bytes
    optional uint32 Size = 2;
};

// Describes document zones in terms of offsets to main doc body
// One TDocZone corresponds to one TArchiveZoneSpan (see ysite/yandex/tarc/tarcface.h)
message TDocZone {
    // Type of document zone, see EArchiveZone enum in ysite/yandex/tarc/tarcface.h
    optional uint32 Type = 1;

    // Offset in TDocArcData::Text to first byte of the zone
    optional uint32 Offset = 2;

    // Size of zone in bytes (may span several sentences)
    optional uint32 Size = 3;

    // Index of first zone sentence in TDocArcData::Sentences
    optional uint32 FirstSentence = 4;
};

// Arbitrary key-value pairs for unstructured storage.
// Keys and values are strings; keys are usually ASCII,
// values are usually UTF-8, but values in csYandex have also been seen.
message TDocInfo {
    optional bytes DocInfoName = 1;
    optional bytes DocInfoValue = 2;
};

// Document data from text archive
message TDocArcData {
    // Document text from archive, UTF-8. No markup. Includes title.
    optional string Text = 1;

    // Document sentences
    repeated TDocSentence Sentences = 2;

    //Document zones
    repeated TDocZone Zones = 3;

    // Site host id
    optional uint32 HostId = 4;

    // Site url id
    optional uint32 UrlId = 5;

    // Timestamp
    optional uint64 Time = 6;

    // Site host name
    optional string HostName = 7;

    // Site url name
    optional string UrlName = 8;

    // Mime type (as string)
    optional string MimeType = 9;

    // Original encoding (see ECharset in util/charset/doccodes.h)
    optional uint32 Charset = 10;

    // Document language (see ELanguage in util/charset/doccodes.h)
    optional uint32 Language = 11;

    // Images global document id. Responsible: krivokon@, images-quality@
    optional string GlobalDocId  = 12;

    // Extracted document title from archive, UTF-8. No markup (SEARCH-1178)
    // Not extracted by default, call SaveDocTitleToProto to fill it.
    optional string Title = 13;

    // Extracted document anchor texts from archive, UTF-8.
    // Not extracted by default, call SaveDocAnchorTextsToProto to fill it.
    repeated string AnchorTexts = 14;

    // Extracted document unstructured properties.
    repeated TDocInfo DocInfos = 15;
};

